controller/Controller.h

```diff
@@ -18,346 +18,61 @@
  * @param socket socket unique ptr.
  * @param op operation to perform
  */
-bool startComponent(std::unique_ptr<org::apache::nifi::minifi::io::Socket> socket, const std::string& component) {
-  return sendSingleCommand(std::move(socket), org::apache::nifi::minifi::c2::Operation::START, component);
-}
+bool startComponent(std::unique_ptr<io::Socket> socket, const std::string& component);
 
 /**
  * Clears a connection queue.
  * @param socket socket unique ptr.
  * @param op operation to perform
  */
-bool clearConnection(std::unique_ptr<org::apache::nifi::minifi::io::Socket> socket, const std::string& connection) {
-  return sendSingleCommand(std::move(socket), org::apache::nifi::minifi::c2::Operation::CLEAR, connection);
-}
+bool clearConnection(std::unique_ptr<io::Socket> socket, const std::string& connection);
 
 /**
  * Updates the flow to the provided file
  */
-int updateFlow(std::unique_ptr<org::apache::nifi::minifi::io::Socket> socket, std::ostream &out, const std::string& file) {
-  socket->initialize();
-  std::vector<uint8_t> data;
-  uint8_t op = org::apache::nifi::minifi::c2::Operation::UPDATE;
-  org::apache::nifi::minifi::io::BufferStream stream;
-  stream.write(&op, 1);
-  stream.write("flow");
-  stream.write(file);
-  if (org::apache::nifi::minifi::io::isError(socket->write(stream.getBuffer()))) {
-    return -1;
-  }
-  // read the response
-  uint8_t resp = 0;
-  socket->read(resp);
-  if (resp == org::apache::nifi::minifi::c2::Operation::DESCRIBE) {
-    uint16_t connections = 0;
-    socket->read(connections);
-    out << connections << " are full" << std::endl;
-    for (int i = 0; i < connections; i++) {
-      std::string fullcomponent;
-      socket->read(fullcomponent);
-      out << fullcomponent << " is full" << std::endl;
-    }
-  }
-  return 0;
-}
+int updateFlow(std::unique_ptr<io::Socket> socket, std::ostream &out, const std::string& file);
 
 /**
  * Lists connections which are full
  * @param socket socket ptr
  */
-int getFullConnections(std::unique_ptr<org::apache::nifi::minifi::io::Socket> socket, std::ostream &out) {
-  socket->initialize();
-  std::vector<uint8_t> data;
-  uint8_t op = org::apache::nifi::minifi::c2::Operation::DESCRIBE;
-  org::apache::nifi::minifi::io::BufferStream stream;
-  stream.write(&op, 1);
-  stream.write("getfull");
-  if (org::apache::nifi::minifi::io::isError(socket->write(stream.getBuffer()))) {
-    return -1;
-  }
-  // read the response
-  uint8_t resp = 0;
-  socket->read(resp);
-  if (resp == org::apache::nifi::minifi::c2::Operation::DESCRIBE) {
-    uint16_t connections = 0;
-    socket->read(connections);
-    out << connections << " are full" << std::endl;
-    for (int i = 0; i < connections; i++) {
-      std::string fullcomponent;
-      socket->read(fullcomponent);
-      out << fullcomponent << " is full" << std::endl;
-    }
-  }
-  return 0;
-}
-
-int getJstacks(std::unique_ptr<org::apache::nifi::minifi::io::Socket> socket, std::ostream &out) {
```

Comment:
It was removed as it was not working, interrupted the process, but did not collect the stacktrace or continue the process so it only freezed the agent. On the other hand this option required C2 agent to be enabled while everything else was implemented on a separate controller socket without the need for C2 to be enabled, so I thought it would be better to remove this option altogether.

Reply:
I checked and the issue was with the `NoOpThreadedRepsitory` where the thread was not running (because it does nothing) and it got stuck when waiting for the thread to continue. I fixed the issue and implemented the jstack command in aeee1f2be56608ee8911112e2a23ac93f6666a37

Reply:
Brilliant, thank you!

I assume the fix is the added check in `ThreadedRepository::getTraces()`?  It could be useful to add a comment there to explain what the check is for.

Reply:
Added comment in 8735105d4cd70f6bbe33fbe43bd17760382a9b1d